contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,193 @@
+# topology aware routing of services
+
+Author: @m1093782566
+
+## Objective
+
+Provide a generic way to provide the local-only service route, say "topology aware routing of service". Local means "same topology level", e.g. same node, same rack, same failure zone, same failure region, same cloud provider etc. To clarify, this proposal is intended to resolve the user case that user always want to visit a local service's backend even though for some reason local service's backend is not available.
```

Comment:
What does "even though for some reason local service ... is not available" mean? Is the goal to provide some sort of anycast or to just return "service not available" in that case...

Reply:
There are two sorts of topology requirements similar to pod Affinity: "hard" and "soft". If "hard" is not met, then just return "service not available". "soft" can be used to calculate how how much preferred the service backend is.